File: py-appscript/branches/py-appscript-0.19.0/lib/appscript/referencerenderer.py
"""referencerenderer -- Generates string representations of appscript references from aem object specifiers.

(C) 2004-2008 HAS"""

from terminology import kProperty, kElement
from aem import Query

######################################################################
# PRIVATE
######################################################################

class _Formatter:
	def __init__(self, appData):
		self._appData = appData
		if self._appData.constructor == 'current':
			self.root = 'app()'
		elif self._appData.constructor == 'path':
			self.root = 'app(%r)' % self._appData.identifier
		else:
			self.root = 'app(%s=%r)' % (self._appData.constructor, self._appData.identifier)
		self.result = ''
	
	def _format(self, val):
		if isinstance(val, Query):
			return renderreference(self._appData, val)
		else:
			return repr(val)
	
	def property(self, code):
		try:
			self.result += '.' + self._appData.referencebycode[kProperty+code][1]
		except KeyError:
			self.result += '.' + self._appData.referencebycode[kElement+code][1]
		return self

	def elements(self, code):
		try:
			self.result += '.' + self._appData.referencebycode[kElement+code][1]
		except KeyError:
			self.result += '.' + self._appData.referencebycode[kProperty+code][1]
		return self
	
	def byname(self, sel):
		self.result += '[%r]' % sel
		return self
	
	byindex = byname
	
	def byid(self, sel):
		self.result += '.ID(%r)' % sel
		return self
	
	def byrange(self, sel1, sel2):
		self.result += '[%s:%s]' %(renderreference(self._appData, sel1), renderreference(self._appData, sel2))
		return self
		
	def byfilter(self, sel):
		self.result += '[%s]' % renderreference(self._appData, sel)
		return self
	
	def previous(self, sel):
		self.result += '.previous(%r)' % self._appData.typebycode[sel]
		return self
	
	def next(self, sel):
		self.result += '.next(%r)' % self._appData.typebycode[sel]
		return self
	
	def __getattr__(self, name):
		if name == 'app':
			self.result += self.root
		elif name == 'NOT':
			self.result = '(%s).NOT' % self.result
		else:
			if name not in ['con', 'its']:
				self.result += '.'
			self.result += name
		return self
	
	def gt(self, sel):
		self.result += ' > %s' % self._format(sel)
		return self
	
	def ge(self, sel):
		self.result += ' >= %s' % self._format(sel)
		return self
	
	def eq(self, sel):
		self.result += ' == %s' % self._format(sel)
		return self
	
	def ne(self, sel):
		self.result += ' != %s' % self._format(sel)
		return self
	
	def lt(self, sel):
		self.result += ' < %s' % self._format(sel)
		return self
	
	def le(self, sel):
		self.result += ' <= %s' % self._format(sel)
		return self
	
	def beginswith(self, sel):
		self.result += '.beginswith(%s)' % self._format(sel)
		return self
	
	def endswith(self, sel):
		self.result += '.endswith(%s)' % self._format(sel)
		return self
	
	def contains(self, sel):
		self.result += '.contains(%s)' % self._format(sel)
		return self
	
	def isin(self, sel):
		self.result += '.isin(%s)' % self._format(sel)
		return self
	
	def AND(self, *operands):
		self.result = '(%s).AND(%s)' % (self.result, ', '.join([self._format(o) for o in operands]))
		return self
		
	def OR(self, *operands):
		self.result = '(%s).OR(%s)' % (self.result, ', '.join([self._format(o) for o in operands]))
		return self


######################################################################
# PUBLIC
######################################################################

def renderreference(appdata, aemreference):
	"""Take an aem reference, e.g.:
	
		app.elements('docu').byindex(1).property('ctxt')
	
	and an AppData instance containing application's location and terminology, and render an appscript-style reference, e.g.:
	
		"app('/Applications/TextEdit.app').documents[1].text"
		
	Used by Reference.__repr__().
	"""
	f = _Formatter(appdata)
	try:
		aemreference.AEM_resolve(f)
	except:
		return '%s.AS_newreference(%r)' % (f.root, aemreference)
	return f.result
	

File: rb-appscript/tags/rb-appscript-0.6.0/src/lib/_appscript/referencerenderer.rb
#
# rb-appscript
#
# referencerenderer -- obtain an appscript-style string representation of an aem reference
#

require "_aem/aemreference"


class ReferenceRenderer
	# Generates string representations of appscript references from aem object specifiers.

	private_class_method :new
	attr_reader :result
	
	def initialize(app_data)
		@_app_data = app_data
		@result = ""
	end
	
	def _format(val)
		if val.is_a?(AEMReference::Query)
			return ReferenceRenderer.render(@_app_data, val)
		else
			return val.inspect
		end
	end
	
	##
	
	def property(code)
		name = @_app_data.reference_by_code.fetch('p'+code) { @_app_data.reference_by_code.fetch('e'+code) }
		@result += ".#{name}"
		return self
	end
	
	def elements(code)
		name = @_app_data.reference_by_code.fetch('e'+code) { @_app_data.reference_by_code.fetch('p'+code) }
		@result += ".#{name}"
		return self
	end
	
	def by_name(name)
		@result += "[#{_format(name)}]"
		return self
	end
	
	def by_index(index)
		@result += "[#{_format(index)}]"
		return self
	end
	
	def by_id(id)
		@result += ".ID(#{_format(id)})"
		return self
	end
	
	def by_range(sel1, sel2)
		@result += "[#{_format(sel1)}, #{_format(sel2)}]"
		return self
	end
	
	def by_filter(sel)
		@result += "[#{_format(sel)}]"
		return self
	end
	
	def previous(sel)
		@result += ".previous(#{_format(@_app_data.type_by_code[sel])})"
		return self
	end
	
	def next(sel)
		@result += ".next(#{_format(@_app_data.type_by_code[sel])})"
		return self
	end
	
	##
	
	def custom_root(value)
		app
		@result += ".AS_new_reference(#{value.inspect})"
		return self
	end
	
	def app
		case @_app_data.constructor
			when :current
				@result = "app.current"
			when :by_path
				@result = "app(#{@_app_data.identifier.inspect})"
		else
			@result = "app.#{@_app_data.constructor}(#{@_app_data.identifier.inspect})"
		end
		return self
	end
	
	def con
		@result = "con"
		return self
	end
	
	def its
		@result = "its"
		return self
	end
	
	##
	
	def beginning
		@result += ".beginning"
		return self
	end
	
	def end
		@result += ".end"
		return self
	end
	
	def before
		@result += ".before"
		return self
	end
	
	def after
		@result += ".after"
		return self
	end
	
	##
	
	def first
		@result += ".first"
		return self
	end
	
	def middle
		@result += ".middle"
		return self
	end
	
	def last
		@result += ".last"
		return self
	end
	
	def any
		@result += ".any"
		return self
	end
	
	##
	
	def gt(val)
		@result += ".gt(#{_format(val)})"
		return self
	end
	
	def ge(val)
		@result += ".ge(#{_format(val)})"
		return self
	end
	
	def eq(val)
		@result += ".eq(#{_format(val)})"
		return self
	end
	
	def ne(val)
		@result += ".ne(#{_format(val)})"
		return self
	end
	
	def lt(val)
		@result += ".lt(#{_format(val)})"
		return self
	end
	
	def le(val)
		@result += ".le(#{_format(val)})"
		return self
	end
	
	def begins_with(val)
		@result += ".begins_with(#{_format(val)})"
		return self
	end
	
	def ends_with(val)
		@result += ".ends_with(#{_format(val)})"
		return self
	end
	
	def contains(val)
		@result += ".contains(#{_format(val)})"
		return self
	end
	
	def is_in(val)
		@result += ".is_in(#{_format(val)})"
		return self
	end
	
	def and(*operands)
		@result += ".and(#{(operands.map { |val| _format(val) }).join(', ')})"
		return self
	end
	
	def or(*operands)
		@result += ".or(#{(operands.map { |val| _format(val) }).join(', ')})"
		return self
	end
	
	def not
		@result += ".not"
		return self
	end
	
	# public
	
	def ReferenceRenderer.render(app_data, aem_reference)
		# Take an aem reference, e.g.:
		#
		#	app.elements('docu').by_index(1).property('ctxt')
		#
		# and an AppData instance containing application's location and terminology, 
		# and render an appscript-style reference string, e.g.:
		#
		#	"AS.app('/Applications/TextEdit.app').documents[1].text"
		#
		# Used by AS::Reference#to_s
		#
		f = new(app_data)
		begin
			aem_reference.AEM_resolve(f)
			return f.result
		rescue
			return "#{new(app_data).app.result}.AS_new_reference(#{aem_reference.inspect})"
		end
	end

end

	
	